operator/pkg/manifest/installer.go

```diff
@@ -267,19 +268,37 @@ func parseKubectlVersion(kubectlStdout string) (*goversion.Version, *goversion.V
 }
 
 // ApplyAll applies all given manifests using kubectl client.
-func ApplyAll(manifests name.ManifestMap, version pkgversion.Version, iop *v1alpha1.IstioOperatorSpec, opts *kubectlcmd.Options) (CompositeOutput, error) {
+func ApplyAll(manifests name.ManifestMap, version pkgversion.Version, iop *v1alpha1.IstioOperatorSpec,
+	opts *kubectlcmd.Options, certDir, clusterName string) (CompositeOutput, error) {
 	scope.Infof("Preparing manifests for these components:")
 	for c := range manifests {
 		scope.Infof("- %s", c)
 	}
 	scope.Infof("Component dependencies tree: \n%s", installTreeString())
-	if _, err := InitK8SRestClient(opts.Kubeconfig, opts.Context); err != nil {
+	restConfig, err := InitK8SRestClient(opts.Kubeconfig, opts.Context)
+	if err != nil {
+		return nil, err
+	}
+	client, err := kubernetes.NewForConfig(restConfig)
+	if err != nil {
 		return nil, err
 	}
 	// Set up the namespace for installation
 	if err := createNamespace(iop.Namespace); err != nil {
 		return nil, err
 	}
+
+	//try to plugIn external certs
+	if certDir != "" {
+		err := cert.PlugInProvisionedCert(client, certDir, iop.Namespace, opts.DryRun)
+		if err != nil {
+			scope.Infof("plugin specified cert and key failed, will generate self signed root cert and two intermediate ca certs")
+		}
+		err = cert.GenerateCertsAndKey(client, iop.Namespace, opts.DryRun, certDir, clusterName)
```

Comment:
Did you forget to put this in the else block? If it is here, it will always be self signed even if we provide certs.

Reply:
when user do not specify `--cert-dir` we will not plug in external cert or generate self-signed certs.

when user specify `--cert-dir`, we will first try to plugin external certs if there is. And if there is we will create create the `cacerts` secret. So in GenerateCertsAndKey we will first check whether the `cacert` secret exists or not. We will generate selfsignedcerts only if `cacert` does not exist:
```_, scrtErr := client.CoreV1().Secrets(ns).Get(context.TODO(), PlugInSecret, metav1.GetOptions{})
	if scrtErr == nil {
		log.Infof(PlugInSecret+"already exists in %v, skip generating self-signed Certs and Keys", ns)
		return nil
	}```